core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java

```diff
@@ -448,6 +434,12 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
 
             if (nodesChanged) {
                 newState.nodes(nodesBuilder);
+                final ClusterState tmpState = newState.build();
+                RoutingAllocation.Result result = routingService.getAllocationService().reroute(tmpState, "node_join");
```

Comment:
We should also do the same in `LocalDiscovery` as we do here.
We can then also remove `getAllocationService` from `RoutingService`.